src/Algebra/Graph/Undirected/UndirectedGraph.hs

```diff
@@ -0,0 +1,935 @@
+rule had a chance to fire.
+
+We could have written the following rule instead:
+
+    RULES "eqIntR" (==) = eqIntR
+
+But that would have to rely on appropriate inlining behaviour of (==) which is
+not under our control. We therefore choose the safe and more explicit path of
+creating our own intermediate functions for guiding rewrite rules when needed.
+-}
+
+instance Functor UndirectedGraph where
+    fmap = fmapR
+
+fmapR :: (a -> b) -> UndirectedGraph a -> UndirectedGraph b
+fmapR f g = bindR g (vertex . f)
+{-# INLINE fmapR #-}
+
+-- | __Note:__ this does not satisfy the usual ring laws; see 'Graph' for more
+-- details.
+instance Num a => Num (UndirectedGraph a) where
+    fromInteger = vertex . fromInteger
+    (+)         = overlay
+    (*)         = connect
+    signum      = const empty
+    abs         = id
+    negate      = id
+
+instance Ord a => Eq (UndirectedGraph a) where
+    (==) = eqR
+
+instance Ord a => Ord (UndirectedGraph a) where
+    compare = ordR
+
+-- TODO: Find a more efficient equality check.
+-- Check if two graphs are equal by converting them to their adjacency maps.
+eqR :: Ord a => UndirectedGraph a -> UndirectedGraph a -> Bool
+eqR x y = toAdjacencyMap x == toAdjacencyMap y
+{-# NOINLINE [1] eqR #-}
+{-# RULES "eqR/Int" eqR = eqIntR #-}
+
+-- Like 'eqR' but specialised for graphs with vertices of type 'Int'.
+eqIntR :: UndirectedGraph Int -> UndirectedGraph Int -> Bool
+eqIntR x y = toAdjacencyIntMap x == toAdjacencyIntMap y
+
+-- TODO: Find a more efficient comparison.
+-- Compare two graphs by converting them to their adjacency maps.
+ordR :: Ord a => UndirectedGraph a -> UndirectedGraph a -> Ordering
+ordR x y = compare (toAdjacencyMap x) (toAdjacencyMap y)
+{-# NOINLINE [1] ordR #-}
+{-# RULES "ordR/Int" ordR = ordIntR #-}
+
+-- Like 'ordR' but specialised for graphs with vertices of type 'Int'.
+ordIntR :: UndirectedGraph Int -> UndirectedGraph Int -> Ordering
+ordIntR x y = compare (toAdjacencyIntMap x) (toAdjacencyIntMap y)
+
+instance Applicative UndirectedGraph where
+    pure  = vertex
+    (<*>) = apR
+
+apR :: UndirectedGraph (a -> b) -> UndirectedGraph a -> UndirectedGraph b
+apR f x = bindR f (<$> x)
+{-# INLINE apR #-}
+
+instance Monad UndirectedGraph where
+    return = pure
+    (>>=)  = bindR
+
+bindR :: UndirectedGraph a -> (a -> UndirectedGraph b) -> UndirectedGraph b
+bindR g f = foldg empty f overlay connect g
+{-# INLINE [0] bindR #-}
+
+instance Alternative UndirectedGraph where
+    empty = empty
+    (<|>) = overlay
+
+instance MonadPlus UndirectedGraph where
+    mzero = empty
+    mplus = overlay
+
+-- | Construct the /empty graph/.
+-- Complexity: /O(1)/ time, memory and size.
+--
+-- @
+-- 'isEmpty'     empty == True
+-- 'hasVertex' x empty == False
+-- 'vertexCount' empty == 0
+-- 'edgeCount'   empty == 0
+-- 'size'        empty == 1
+-- @
+empty :: UndirectedGraph a
+empty = UndirectedGraph G.Empty
```

Comment:
Most of the functions in the API can be implemented using type-safe coercion, e.g. `empty = coerce G.empty`.

Reply:
Got it! :)

Reply:
How can I resolve this error?

![image](https://user-images.githubusercontent.com/20261421/61046436-093f0f80-a3d5-11e9-9952-edf98a326f85.png)